libraries/standard/mqtt/cbmc/sources/mqtt_cbmc_state.c

```diff
@@ -18,13 +18,139 @@
  * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  */
-#include <stddef.h>
 #include <stdint.h>
 #include <stdlib.h>
 #include "mqtt_cbmc_state.h"
+#include "mqtt.h"
 
 void * mallocCanFail( size_t size )
 {
     __CPROVER_assert( size < CBMC_MAX_OBJECT_SIZE, "mallocCanFail size is too big" );
     return nondet_bool() ? NULL : malloc( size );
 }
+
+MQTTPacketInfo_t * allocateMqttPacketInfo( MQTTPacketInfo_t * pPacketInfo )
+{
+    if( pPacketInfo == NULL )
+    {
+        pPacketInfo = mallocCanFail( sizeof( MQTTPacketInfo_t ) );
+    }
+
+    if( pPacketInfo != NULL )
+    {
+        __CPROVER_assume( pPacketInfo->remainingLength < CBMC_MAX_OBJECT_SIZE );
+        pPacketInfo->pRemainingData = mallocCanFail( pPacketInfo->remainingLength );
+    }
+
+    return pPacketInfo;
+}
+
+bool isValidMqttPacketInfo( const MQTTPacketInfo_t * pPacketInfo )
+{
+    bool isValid = true;
+
+    if( pPacketInfo != NULL )
+    {
+        isValid = pPacketInfo->remainingLength < CBMC_MAX_OBJECT_SIZE;
+    }
+
+    return isValid;
+}
```

Comment:
This means that packetInfo == NULL is a valid packetInfo.  I kind of like this, but CBMC proof projects are inconsistent from project to project about whether valid means nonnull.  I like this style if you can make it work, but it needs to be used consistently.

Reply:
Yes a NULL packet info is a valid packet info.

Reply:
ELI5, please.

Reply:
@markrtuttle  All isvalid routines in this file have this same style.

@dan4thewin A NULL packet info is valid because we want the proof, that calls this fuction, to accept a NULL packet info to pass into the function to be proved. This allows us to get full coverage and make sure the the function returns an error for NULL parameters.

Reply:
I understand that logic, but I would have thought a simple NULL check would be in a unit test.  Is that an integral part of a this proof?